src/createExplorer.js

```diff
@@ -39,43 +57,66 @@ module.exports = function createExplorer(options) {
         return fileCache.get(absoluteConfigPath);
       }
 
-      const load =
-        path.basename(absoluteConfigPath) === 'package.json'
-          ? () => loadPackageProp(path.dirname(absoluteConfigPath), options)
-          : () => loadDefinedFile(absoluteConfigPath, options);
+      let load;
+      if (path.basename(absoluteConfigPath) === 'package.json' && packageProp) {
+        load = () =>
+          loadPackageProp(path.dirname(absoluteConfigPath), {
+            packageProp,
+            sync: options.sync,
+          });
+      } else {
+        load = () =>
+          loadDefinedFile(absoluteConfigPath, {
+            sync: options.sync,
+            format: options.format,
+          });
+      }
 
-      const result = !options.sync ? load().then(transform) : transform(load());
+      const loadResult = load();
+      const result =
+        loadResult instanceof Promise
+          ? loadResult.then(transform)
+          : transform(loadResult);
       if (fileCache) fileCache.set(absoluteConfigPath, result);
       return result;
     }
 
     if (!searchPath) return !options.sync ? Promise.resolve(null) : null;
 
     const absoluteSearchPath = path.resolve(process.cwd(), searchPath);
-    const searchPathDir = resolveDir(absoluteSearchPath, options.sync);
+    const searchPathDir = getDirectory(absoluteSearchPath, options.sync);
 
-    return !options.sync
+    return searchPathDir instanceof Promise
```

Comment:
Is this something flow forces us to do? Could we use `is-promise` for this?

Reply:
Oh, we can use `isPromise`, yes, didn't think of that. 

For all of these functions that return `Promise<string> | string`, Flow doesn't have a good way of knowing if it should check of a string or a Promise. So if you just attach `.then()` to the result, it will complain because that could a string, not a Promise. This is essentially the same check (will be true in all the same situations), but gives Flow an opportunity to check what's going on.

Reply:
Tried this out. `isPromise` doesn't provide the same type safety, and I think we don't need it, actually, because we are using real Promises. I can replace other usage of `isPromise` with an `instanceof` check.

Reply:
Okay. But IMHO we should avoid `instanceof` checks. Not sure if this could affect us, but with various implementations of promise, an instance check might not be reliable. See https://github.com/petkaantonov/bluebird/issues/628. And the recommended method is to check the presense of `.then` - https://stackoverflow.com/questions/27746304/how-do-i-tell-if-an-object-is-a-promise.

Reply:
If you'd like to work on getting `isPromise` into these situations, go for it. I doubt we have to worry about this since we are using real Promises — and tests should fail if we aren't. But I totally understand your point.